### PR TITLE
Add AzureOIDC provider

### PR DESCRIPTION
**Pull request by thereallukl**
## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/azure_oidc.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	Groups []string
+}
+
+// NewAzureOIDCProvider initiates a new AzureOIDCProvider
+func NewAzureOIDCProvider(p *ProviderData) *AzureOIDCProvider {
+	provider := AzureOIDCProvider{
+		OIDCProvider: NewOIDCProvider(p),
+	}
+	provider.ProviderName = "Azure OpenID Connect"
+	provider.ApprovalPrompt = "consent"
```
Comment (eerichmond) [inline providers/azure_oidc.go]: Can you make the ApprovalPrompt come from the input param instead of hardcoding as "consent"

File: providers/azure_oidc.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	p.Scope = "openid email profile"
```
Comment (eerichmond) [inline providers/azure_oidc.go]: Scope is being hardcoded here as well as set inside of NewOIDCProvider.

File: providers/azure_oidc.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	for _, exemptedEmail := range p.ExemptedEmails {
+		if session.Email == exemptedEmail {
+			return true
+		}
+	}
+	var claims AzureOIDCClaims
+	err := getClaimFromToken(session.IDToken, &claims)
+	if err != nil {
+		logger.Printf("error: failed to parse IDToken")
```
Comment (eerichmond) [inline providers/azure_oidc.go]: Can you include the original err in the logger.Printf?

File: main.go [hunk 4]
```diff
@@ -25,6 +25,8 @@ func main() {
 	skipAuthRegex := StringArray{}
 	jwtIssuers := StringArray{}
 	googleGroups := StringArray{}
+	azurePermittedGroups := StringArray{}
```
Comment (eerichmond) [inline main.go]: Maybe shorten this to just azureGroups to match googleGroups above it

**Comment by eerichmond** [comment]
I would love to see this merged into pusher/oauth2_proxy. Let me know if you want me to contribute to the message that goes into the pull request.

**Comment by eerichmond** [comment]
Have you given up on this fork? Should I fork it and maintain it myself?

**Comment by andymelichar** [comment]
Trying this out for myself... how are you handling the email_verified from Azure? It doesn't look like it's part of the default payload and so I'm getting back "Error redeeming code during OAuth2 callback: email <myemailaddress> not listed as verified"
